src/plot.cpp

```diff
@@ -1,10 +1,991 @@
+      }
+
+    } else {
+      std::stringstream err_msg;
+      err_msg << "Mutliple masks specified in plot " << id;
+      fatal_error(err_msg);
+    }
+  }
+}
+
+Plot::Plot(pugi::xml_node plot_node):
+index_meshlines_mesh(-1)
+{
+  set_id(plot_node);
+  set_type(plot_node);
+  set_output_path(plot_node);
+  set_bg_color(plot_node);
+  set_basis(plot_node);
+  set_origin(plot_node);
+  set_width(plot_node);
+  set_universe(plot_node);
+  set_default_colors(plot_node);
+  set_user_colors(plot_node);
+  set_meshlines(plot_node);
+  set_mask(plot_node);
+} // End Plot constructor
+
+//==============================================================================
+// POSITION_RGB computes the red/green/blue values for a given plot with the
+// current particle's position
+//==============================================================================
+
+
+void position_rgb(Particle p, Plot pl, RGBColor &rgb, int &id)
+{
+  p.n_coord = 1;
+
+  bool found_cell = find_cell(&p, 0);
+
+  int j = p.n_coord - 1;
+
+  if (settings::check_overlaps) {check_cell_overlap(&p);}
+
+  // Set coordinate level if specified
+  if (pl.level >= 0) {j = pl.level + 1;}
+
+  if (!found_cell) {
+    // If no cell, revert to default color
+    rgb = pl.not_found;
+    id = -1;
+  } else {
+    if (plot_color_by::mats == pl.color_by) {
+      // Assign color based on material
+      Cell* c = cells[p.coord[j].cell];
+      if (c->type_ == FILL_UNIVERSE) {
+        // If we stopped on a middle universe level, treat as if not found
+        rgb = pl.not_found;
+        id = -1;
+      } else if (p.material == MATERIAL_VOID) {
+        // By default, color void cells white
+        rgb = WHITE;
+        id = -1;
+      } else {
+        rgb = pl.colors[p.material - 1];
+        id = materials[p.material - 1]->id_;
+      }
+
+    } else if (plot_color_by::cells == pl.color_by) {
+      // Assign color based on cell
+      rgb = pl.colors[p.coord[j].cell];
+      id = cells[p.coord[j].cell]->id_;
+    } else {
+      rgb = NULLRGB;
+      id = -1;
+    }
+
+  } // endif found_cell
+}
+
+//==============================================================================
+// OUTPUT_PPM writes out a previously generated image to a PPM file
+//==============================================================================
+
+void output_ppm(Plot pl, const ImageData &data)
+{
+  // Open PPM file for writing
+  std::string fname = pl.path_plot;
+  fname = strtrim(fname);
+  std::ofstream of;
+
+  of.open(fname);
+
+  // Write header
+  of << "P6" << std::endl;
+  of << pl.pixels[0] << " " << pl.pixels[1] << std::endl;
+  of << "255" << std::endl;
+  of.close();
+
+  of.open(fname, std::ios::binary | std::ios::app);
+  // Write color for each pixel
+  for (int y = 0; y < pl.pixels[1]; y++) {
+    for (int x = 0; x < pl.pixels[0]; x++) {
+      RGBColor rgb = data(x,y);
+      of.write((char*)&rgb[RED], 1);
+      of.write((char*)&rgb[GREEN], 1);
+      of.write((char*)&rgb[BLUE], 1);
```

Comment:
`static_cast<char*>(...)` should work here, right?

Reply:
It might work, but if you're writing binary data to a file (and it's required to be `char*`), the meaning of `reinterpret_cast` is usually "don't change the bit pattern" which seems appropriate.

Reply:
Oh wait, does this block of code only work for little-Endian ints?  As best I can tell, you're just writing out the first byte of a 4-byte int.

Reply:
Yeah I think you're right @smharper. Your suggestion to use `char` for the `RGBColor` array should take care of that.